### DISCUSSION TITLE
CAD Framework Evaluation

### DISCUSSION THREAD

**Discussion by DJGosnell** [ppy/osu-framework#5037] [Q&A] [answered]
I am evaluating this framework to determine if this is a good fit or can be modified to be a good fit for my project.  I am looking at using this framework for cross platform, CAD interface. I have been doing research inside the framework but I have yet to find any kind of low level systems to draw lines on a Drawable without some kind of modification to the rendering pipeline.  

Here are my questions about what the framework can and can't do at this point in time.

- Provide low level access to the rendering pipeline to draw many lines (10s of thousands), with or without hit detection.
- Built-in panning and zooming of Drawable content.
- Ability to disable the requirement of the BASS library.

Your help is appreciated as I like a lot of the features of this framework and I would like to see it more widely used.

**Comment by bdach** [comment, marked as answer]
> Provide low level access to the rendering pipeline to draw many lines (10s of thousands), with or without hit detection.

Somewhat doable via drawables with custom draw nodes - you'd have to set up your own vertex batches ([example from game](https://github.com/ppy/osu/blob/45db99171e9e43904b10405fc77dd104451d6eff/osu.Game/Graphics/Backgrounds/Triangles.cs#L271-L316)). If you're looking at even lower level then you'd have to issue your own GL calls.

Hit detection you'd have to roll your own.

> Built-in panning and zooming of Drawable content.

This depends on what you're looking for. Game already has [zoomable containers](https://github.com/ppy/osu/blob/45db99171e9e43904b10405fc77dd104451d6eff/osu.Game/Screens/Edit/Compose/Components/Timeline/ZoomableScrollContainer.cs) but with a CAD solution I presume you're looking for essentially a vector graphics editor (limitless zoom, etc.) Probably would be a fair amount of work involved to make that work fast (i.e. limit rendered content to visible extents, etc.)

> Ability to disable the requirement of the BASS library.

Pretty much non-negotiable. This is a framework for a rhythm game first and foremost. I'd wager that nobody from the team or the regular contributors would be willing to spend time on tearing audio support out for one third-party consumer.

---

In general I'd say that there is a high likelihood that this framework is a poor fit for what you want to be doing. If you wish for low-level access then honestly raw OpenGL + SDL combined with some UI framework (dear imgui/qt?) may be the best fit for you.

**Reply by peppy** [reply]
Pretty good coverage.

I think based on your requirements you may be looking for something different. While the framework is written in a way BASS could be easily removed if you choose to fork and do so, as your low level draw requirements increased the usefulness you may get from osu!framework will drop.

As some pointers, take a look at silk.net, veldrid (both very low level if you want to make your own UI system) or MAUI / Flutter if you want GPU accelerated non-electron prebuilt UI frameworks.

**Reply by DJGosnell** [reply]
It was looking that way the more I evaluated the system.  I very much liked the idea of mixing lower level drawing and higher level UI interactions, but I think the more I look at it, it would require a good amount of work to add a separate drawing sub-system into the rendering pipeline.

Thank you for your time in helping my evaluation.
